### PR TITLE
R2 DH

### PR DESCRIPTION
**Pull request by MohamedLaghdafHABIBOULLAH**

### CODE REVIEW

File: examples/demo-bpdn-constr-L0.jl [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+function demo_bpdn_constr(compound = 1)
+  model, nls_model, sol = bpdn_model(compound, bounds = true)
+  f = LSR1Model(model)
+  λ = norm(grad(model, zeros(model.meta.nvar)), Inf) / 10
+  demo_solver(f, nls_model, sol, NormL0(λ), NormLinf(1.0))
+  #λ = norm(grad(model, zeros(model.meta.nvar)), Inf) / 3
+  #demo_solver(f, nls_model, sol, NormL1(λ), NormLinf(1.0), "l1-linf")
+end
+
+demo_bpdn_constr()
```
Comment (dpo) [inline examples/demo-bpdn-constr-L0.jl:69]: Please simply add your solver to the existing demos instead of creating a new demo.

File: examples/demo-bpdn-constr.jl [hunk 2]
```diff
@@ -59,4 +59,4 @@ function demo_bpdn_constr(compound = 1)
   demo_solver(f, nls_model, sol, NormL1(λ), NormLinf(1.0), "l1-linf")
 end
 
-demo_bpdn_constr()
+demo_bpdn_constr()
```
Comment (dpo) [inline examples/demo-bpdn-constr.jl:62]: Please undo this; there is no change here.

File: examples/demo-bpdn.jl [hunk 3]
```diff
@@ -38,4 +38,4 @@ function demo_bpdn(compound = 1)
   demo_solver(f, sol, NormL1(λ), NormLinf(1.0), "l1-linf")
 end
 
-demo_bpdn()
+demo_bpdn()
```
Comment (dpo) [inline examples/demo-bpdn.jl:41]: Same here.

**Review by dpo — CHANGES_REQUESTED** [review]
Thanks @MohamedLaghdafHABIBOULLAH. Please add your solver to the existing demos.

**Comment by dpo** [comment]
@MohamedLaghdafHABIBOULLAH Please update this.

File: benchmarks/tables/nnmf-table.jl [hunk 4]
```diff
@@ -3,10 +3,10 @@ include("regulopt-tables.jl")
 Random.seed!(1234)
 m, n, k = 100, 50, 5
 model, A, selected = nnmf_model(m, n, k)
-f = LSR1Model(model)
+f = LBFGSModel(model)
```
Comment (geoffroyleconte) [inline benchmarks/tables/nnmf-table.jl:6]: Why did you change the model here?

File: src/R2DH_alg.jl [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+      push!(Dk, s, ∇fk - ∇fk⁻) # update QN operator
+      DkNorm = norm(Dk.d, Inf) 
+      ∇fk⁻ .= ∇fk
+    end
+
+    if ρk < η1 || ρk == Inf
+      σk = σk * γ
+    end
+
+    σkdk = sum ? Dk.d .+ σk : Dk.d  .* σk
```
Comment (geoffroyleconte) [inline src/R2DH_alg.jl:241]: Same here, you can use `.=` or `@.` for the whole line to avoid allocating a vector every time. If this does not work with the `? ... : ...` synthax, you can split it with a `if .. else`

**Review by geoffroyleconte — CHANGES_REQUESTED** [review]
Thanks for his PR! Here are some comments.

File: src/R2DH_alg.jl (hunk 5 above)
Comment (geoffroyleconte) [reply src/R2DH_alg.jl:241]: Also maybe using `sum` as a bool here is a bit ambiguous here since this is also the name of a function? Could you find another name?

File: benchmarks/tables/nnmf-table.jl (hunk 4 above)
Comment (MohamedLaghdafHABIBOULLAH) [reply benchmarks/tables/nnmf-table.jl:6]: because, the Bk of LSR1model appears to be non positive hence it gives errors in R2DH

File: src/R2DH_alg.jl [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+
+  optimal = false
+  tired = maxIter > 0 && k ≥ maxIter || elapsed_time > maxTime
+
+  while !(optimal || tired)
+    k = k + 1
+    elapsed_time = time() - start_time
+    Fobj_hist[k] = fk
+    Hobj_hist[k] = hk
+    σkdk = max.(σkdk, eps())
```
[the remark this answers is not in the document]
Comment (MohamedLaghdafHABIBOULLAH) [reply src/R2DH_alg.jl:178]: Yeah it is a vector

File: src/R2DH_alg.jl (hunk 5 above)
Comment (MohamedLaghdafHABIBOULLAH) [reply src/R2DH_alg.jl:241]: Ok thank you

File: src/R2DH_alg.jl [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+    min f(x) + h(x)
+
+where f: ℝⁿ → ℝ has a Lipschitz-continuous gradient, and h: ℝⁿ → ℝ is
+lower semi-continuous, proper and prox-bounded.
+
+About each iterate xₖ, a step sₖ is computed as a solution of
+
+    min  φ(s; xₖ) + ψ(s; xₖ)
+
+where φ(s ; xₖ) = f(xₖ) + ∇f(xₖ)ᵀs + ½ sᵀ (σₖ+Dₖ) s (if summation = true) and φ(s ; xₖ) = f(xₖ) + ∇f(xₖ)ᵀs + ½ sᵀ (σₖ+Dₖ) s (if summation = true) is a quadratic approximation of f about xₖ,
```
Comment (geoffroyleconte) [inline src/R2DH_alg.jl:18]: You said 2 times if summation = true.
Add the single quote  to indicate that this is some code.
You can preview the Docs locally by running the `makedocs` function in docs/make.jl of this repo.

File: src/R2DH_alg.jl [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+* `h`: a regularizer such as those defined in ProximalOperators
+* `options::ROSolverOptions`: a structure containing algorithmic parameters
+* `x0::AbstractVector`: an initial guess (in the second calling form)
+
+### Keyword Arguments
+
+* `x0::AbstractVector`: an initial guess (in the first calling form: default = `nlp.meta.x0`)
+* `selected::AbstractVector{<:Integer}`: (default `1:length(x0)`).
+* `Bk`: initial diagonal Hessian approximation (default: `(one(R) / options.ν) * I`).
+* `summation`: boolean in order to choose between the two versions of R2DH (default : `true``).
```
Comment (geoffroyleconte) [inline src/R2DH_alg.jl:34]:
```suggestion
* `summation`: boolean used to choose between the two versions of R2DH (see above, default : `true`).
```

**Review by geoffroyleconte — CHANGES_REQUESTED** [review]

File: src/TRDH_alg.jl [hunk 9]
```diff
@@ -120,6 +120,8 @@ function TRDH(
   β = options.β
   spectral = options.spectral
   psb = options.psb
+  andrei = options.andrei
+  wolk = options.wolk
```
Comment (dpo) [inline src/TRDH_alg.jl:124]: Let's use `wolkowicz` instead of `wolk`.
